### PR TITLE
Modernized AlcaBeamMonitor

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

- Avoid segmentation fault when a job has no LuminosityBlocks
- Added fillDescriptions
- Added esConsumes call

#### PR validation:

Code compiles and the new unit test passes.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
@smuzaffar @makortel this should fix the seg faults seen in the ASAN IBs.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
I do not know why the histograms are different than the base-line. I checked and my addition of 
```C++
if (processedLumis_.empty()) {
    return;
}```

Is not preventing the function from executing to its end for workflow 4.53 (which is the one I just ran). When I look at the results from the harvesting step I do, indeed, see that histograms in `AlcaBeamMonitor/Debug` directory are empty, unlike what the base-line is reporting.

**Comment by Dr15Jones** [comment]
I found the problem. The `esConsumes` didn't say the data was being requested at being LuminosityBlock time. That caused an exception when the `getHandle` was made. However, the `getHandle` was embedded in a `try {}catch` block and that block just output the exception message to `LogInfo` but info messages are suppressed in the standard configuration.

The ultimate take-away is one should never catch exceptions from data requests since problems unrelated to missing data can occur.

**Comment by Dr15Jones** [comment]
please test

**Comment by jfernan2** [comment]
What about the build errors? https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b1948b/9117/newer-than-target.log

**Comment by Dr15Jones** [comment]
@smuzaffar what’s up with the messages that imply build order problems?

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
Let’s try without the extra pull requests.

**Comment by smuzaffar** [comment]
@Dr15Jones , the build rules errors has been fixed for next IB. Please ignore that error for this PR.

**Comment by Dr15Jones** [comment]
Just a reminder, this will avoid the 350 segmentation faults seen in the ASAN IB.

File: RecoVertex/BeamSpotProducer/src/BeamFitter.cc [hunk 1]
```diff
@@ -221,6 +221,43 @@ BeamFitter::~BeamFitter() {
   delete MyPVFitter;
 }
 
+void BeamFitter::fillDescription(edm::ParameterSetDescription &iDesc) {
+  edm::ParameterSetDescription beamFitter;
+
+  beamFitter.addUntracked<bool>("Debug");
+  beamFitter.addUntracked<edm::InputTag>("TrackCollection");
+  iDesc.addUntracked<edm::InputTag>("primaryVertex", edm::InputTag("offlinePrimaryVertices"));
```
Comment (jpata) [inline RecoVertex/BeamSpotProducer/src/BeamFitter.cc:229]: If the default values for these parameters are defined here, would it make sense to remove them from the constructor: https://github.com/cms-sw/cmssw/blob/706cb39da184559708db0e6e9899f49995f65eb0/RecoVertex/BeamSpotProducer/src/BeamFitter.cc#L44 ?

Comment (jpata) [reply RecoVertex/BeamSpotProducer/src/BeamFitter.cc:229]: It seems like the main python definition of BeamFitter is here [1] (with some alternate defitions in neighboring files) - do we want to reduce the duplication of the default parameters somewhat between the fillDescriptions and the cff?

[1] https://github.com/cms-sw/cmssw/blob/02d4198c0b6615287fd88e9a8ff650aea994412e/DQM/BeamMonitor/python/BeamMonitor_cff.py#L14

Comment (Dr15Jones) [reply RecoVertex/BeamSpotProducer/src/BeamFitter.cc:229]: I’d prefer to have that done in a separate pull request so we can get the bug fix in.

**Comment by jfernan2** [comment]
+1

**Comment by Dr15Jones** [comment]
type bug

**Comment by jpata** [comment]
@Dr15Jones what do you mean by the type bug?

**Comment by Dr15Jones** [comment]
@jpata 
>  what do you mean by the type bug?

This pull request gets marked as a bug fix. Right now we are getting segmentation faults in the ASAN IB. This pull request fixes that. (I admit I did more than the bare minimum to fix the bug).

File: DQM/BeamMonitor/plugins/AlcaBeamMonitor.cc [hunk 2]
```diff
@@ -31,24 +29,22 @@ using namespace reco;
...
 //----------------------------------------------------------------------------------------------------------------------
 AlcaBeamMonitor::AlcaBeamMonitor(const ParameterSet& ps)
-    : parameters_(ps),
-      monitorName_(parameters_.getUntrackedParameter<string>("MonitorName", "YourSubsystemName")),
-      primaryVertexLabel_(
-          consumes<VertexCollection>(parameters_.getUntrackedParameter<InputTag>("PrimaryVertexLabel"))),
-      trackLabel_(consumes<reco::TrackCollection>(parameters_.getUntrackedParameter<InputTag>("TrackLabel"))),
-      scalerLabel_(consumes<BeamSpot>(parameters_.getUntrackedParameter<InputTag>("ScalerLabel"))),
-      beamSpotLabel_(parameters_.getUntrackedParameter<InputTag>("BeamSpotLabel")),
+    : monitorName_(ps.getUntrackedParameter<string>("MonitorName")),
```
Comment (christopheralanwest) [inline DQM/BeamMonitor/plugins/AlcaBeamMonitor.cc:32]: Could you explain why you have removed the second parameter `"YourSubsystemName"` here but not in `AlcaBeamMonitor::fillDescriptions`?

Comment (Dr15Jones) [reply DQM/BeamMonitor/plugins/AlcaBeamMonitor.cc:32]: Having it in both the `getUntrackedParameter` call and in `fillDescriptions` would be redundant. If the parameter is missing, the `fillDescriptions` call will inject the `"MonitorName"` string parameter into the module's PSet using the default `"YourSubsystemName"`. Therefore during the constructor call the parameter `"MonitorName"` is guaranteed to exist so no default value is needed for the `getUntrackedParameter`. So the change I made has exactly the same ultimate behavior as the original code.

In general, the framework group highly encourages everyone to stop specifying default values in `getUntrackedParameter` and instead use `fillDescriptions`.

**Comment by jpata** [comment]
+reconstruction

- fixes a bug in AlcaBeamMonitor in case of no LuminosityBlocks
- also moves parameters to fillDescriptions
- some residual cleanup with respect to the python config would be good as a follow-up
- no reco changes expected or observed

**Comment by christopheralanwest** [comment]
+alca

**Comment by silviodonato** [comment]
+1

**Pull request merged**